File: 37.解数独.py
#
# @lc app=leetcode.cn id=37 lang=python3
#
# [37] 解数独
#

# @lc code=start
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        list_row = [{str(num):False for num in range(1, 10)} for _ in range(9)]
        list_col = [{str(num):False for num in range(1, 10)} for _ in range(9)]
        list_square = [{str(num): False for num in range(1, 10)} for _ in range(9)]
        blank_space = list()

        for i in range (9):
            for j in range(9):
                if board[i][j] in '123456789':
                    k = (i // 3) * 3 + j // 3
                    list_row[i][board[i][j]] = True
                    list_col[j][board[i][j]] = True
                    list_square[k][board[i][j]] = True
                else:
                    blank_space.append((i,j))

        def solve():
            if not blank_space:
                return True
            
            i, j = blank_space.pop()
            k = (i // 3) * 3 + j // 3
            for n in '123456789':
                if list_row[i][n] or list_col[j][n] or list_square[k][n]:
                    continue
                list_row[i][n] = True
                list_col[j][n] = True
                list_square[k][n] = True
                board[i][j] = n
                if solve():
                    return True
                list_row[i][n] = False
                list_col[j][n] = False
                list_square[k][n] = False
                board[i][j] = '.'
            blank_space.append((i, j))
            return False   

        solve()
        


# @lc code=end

File: 37.解数独.java
/*
 * @lc app=leetcode.cn id=37 lang=java
 *
 * [37] 解数独
 */

// @lc code=start
class Solution {
    public void solveSudoku(char[][] board) {
        int[][] rows = new int[9][9];
        int[][] cols = new int[9][9];
        int[][] sbox = new int[9][9];

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                int num = board[i][j] - '1';
                int idx_box = (i/3)*3 + j/3;
                rows[i][num] = 1;
                cols[j][num] = 1;
                sbox[idx_box][num] = 1;
            }
        }

        for (int i = 0; i < 9; i++) {
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') {
                    if (fillSudoku(board, rows, cols, sbox, i, j)) return;
                    
                }
            }
        }
    }

    private boolean fillSudoku(char[][] board, int[][] rows, int[][] cols, int[][] sbox, int i, int j) {
        for (int k = 0; k < 9; k++) {
            if (rows[i][k] == 1) continue;
            if (cols[j][k] == 1) continue;
            int idx_box = (i/3) * 3 + j / 3;
            if (sbox[idx_box][k] == 1) continue;

            board[i][j] = (char)('1' + k);
            rows[i][k] = 1;
            cols[j][k] = 1;
            sbox[idx_box][k] = 1;

            int nextI = -1, nextJ = -1;
            for (int newI = 0; newI < 9; newI++) {
                boolean find = false;
                for (int newJ = 0; newJ < 9; newJ++) {
                    if (board[newI][newJ] == '.') {
                        nextI = newI;
                        nextJ = newJ;
                        find = true;
                        break;
                    }
                }
                if (find) break;
            }
            if (nextI == -1) return true;
            if(fillSudoku(board, rows, cols, sbox, nextI, nextJ)) return true;
            board[i][j] = '.';
            rows[i][k] = 0;
            cols[j][k] = 0;
            sbox[idx_box][k] = 0;
        }

        return false;
    }
}
// @lc code=end

